[AMDGPU] Extend promotion of alloca to vectors

perlfu opened the pull request:
* Add multi dimensional array support
* Make maximum vector size tunable
* Make ratio of VGPRs used for vector promotion tunable

perlfu committed:
[AMDGPU] Extend promotion of alloca to vectors

* Add multi dimensional array support
* Make maximum vector size tunable
* Make ratio of VGPRs used for vector promotion tunable

Pierre-vh commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -1546,180 +1546,184 @@ The AMDGPU backend supports the following LLVM IR attributes.
-                                                  temporarily raise the wave priority at the start of the shader function
-                                                  until its last VMEM instructions to allow younger waves to issue their VMEM
-                                                  instructions as well.
+     =============================================== ==========================================================
```
What's this change?

perlfu replied:
Reformat table to accommodate the wider attributes.  Given the doc build failed, I probably didn't make it wide enough.

shiltian replied:
could you do a NFC commit for the reformatting first?

Pierre-vh replied:
Yes please make another diff for the widening. I know it's a bit silly but right now it's impossible to see what changed in the table because the diff shows it all in red.

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUPromoteAlloca.cpp:
```diff
@@ -310,12 +322,17 @@ bool AMDGPUPromoteAllocaImpl::run(Function &F, bool PromoteToLDS) {
+  const unsigned VGPRRatio =
+      PromoteAllocaToVectorVGPRRatio.getNumOccurrences()
+          ? PromoteAllocaToVectorVGPRRatio
+          : F.getFnAttributeAsParsedInteger(
```
Can you put this in a variable to help readability?

perlfu replied:
Sorry, I don't understand -- this is all variables?

Given Matt's comment about pass parameters, this will likely move to being a variable in the pass class definition.

Pierre-vh replied:
No, I mean outlining `F.getFnAttributeAsParsedInteger(...)` into a variable and reusing the variable after the colon, instead of having  3 lines of function call nested inside a ternary

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUPromoteAlloca.cpp:
```diff
@@ -735,23 +768,48 @@ bool AMDGPUPromoteAllocaImpl::tryPromoteAllocaToVector(AllocaInst &Alloca) {
+  const unsigned MaxElements =
+      PromoteAllocaToVectorMaxElements.getNumOccurrences()
+          ? PromoteAllocaToVectorMaxElements
+          : Alloca.getParent()->getParent()->getFnAttributeAsParsedInteger(
```
same here

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUPromoteAlloca.cpp:
```diff
@@ -66,6 +66,18 @@ static cl::opt<unsigned> PromoteAllocaToVectorLimit(
     cl::desc("Maximum byte size to consider promote alloca to vector"),
     cl::init(0));
 
+static cl::opt<unsigned> PromoteAllocaToVectorMaxElements(
+    "amdgpu-promote-alloca-to-vector-max-elements",
+    cl::desc("Maximum vector size (in elements) to use when promoting alloca"),
+    cl::init(16));
```
Do we expect end users to use these options?

perlfu replied:
> Elements seems like a strange way to express this.

Element count is how the limit is currently defined in the code.  I agree in terms of 32-bit words (registers) would make more sense.  I'll change to this model, but it does mean this patch will not preserve the existing limit so some edge case promotion will change.

> Do we expect end users to use these options?

Graphics front end will use these for shader tuning, which is why they are accessible via function attributes as well.

shiltian replied:
Okay. The reason I was asking is, if we expect any uses from users, we need to expose them as an option instead of a pass option, but based on your description, we don't expect from end users, and compiler front end can tune it via function attributes.

perlfu committed:
Address reviewer comments:
- Move options into pass attributes
- Change vector size limit from max elements to max 32b registers
- Add tests for i16, float and ptr in multi-dimensional arrays

perlfu commented on llvm/test/CodeGen/AMDGPU/promote-alloca-multidim.ll:
```diff
@@ -0,0 +1,292 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -S -mtriple=amdgcn-amd-amdhsa -passes=amdgpu-promote-alloca -disable-promote-alloca-to-lds=1 < %s | FileCheck %s
+
+define amdgpu_kernel void @i32_2d_load_store(ptr %out) {
```
Done.

perlfu commented:
Note: I have changed the max vector size to be based on number of 32b registers.  With the current value of 16, this mean some alloca which were previously promoted are no longer promoted.
See promote-alloca-subvecs.ll and vector-alloca-limits.ll where I had to increase the limit for the tests to pass.
Alternatively we could raise the default to 32 to allow 16 x i64/double/ptr to be promoted as previously, with the effect that 32 x i32/float would now be promoted.

perlfu commented:
Ping

perlfu committed:
- Hide generated undefs in lit test to avoid deprecation warning
